core/src/main/java/org/elasticsearch/tribe/TribeService.java

```diff
@@ -463,6 +487,44 @@ private ClusterState applyUpdate(ClusterState currentState, ClusterChangedEvent
                 }
             }
 
+            Set<String> changedCustomMetaDataTypeSet = tasks.stream()
+                    .map(ClusterChangedEvent::changedCustomMetaDataSet)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toSet());
+            final List<Node> tribeNodes = TribeService.this.nodes;
+            Map<String, MetaData.Custom> mergedCustomMetaDataMap = mergeChangedCustomMetaData(changedCustomMetaDataTypeSet,
+                    customMetaDataType -> {
+                        List<MetaData.Custom> tribeCustomMetaDataList = new ArrayList<>(tribeNodes.size());
+                        for (Node tribeNode : tribeNodes) {
+                            String currentTribeName = TRIBE_NAME_SETTING.get(tribeNode.settings());
+                            final MetaData.Custom custom;
+                            if (currentTribeName.equals(tribeName)) {
+                                // current tribe node
+                                custom = tribeState.metaData().custom(customMetaDataType);
+                            } else {
+                                ClusterState currentTribeState = tribeNode.injector().getInstance(ClusterService.class).state();
+                                custom = currentTribeState.metaData().custom(customMetaDataType);
+                            }
+                            if (custom != null && custom instanceof MergableCustomMetaData) {
+                                tribeCustomMetaDataList.add(custom);
+                            }
+                        }
+                        return tribeCustomMetaDataList;
+            });
+            for (String changedCustomMetaDataType : changedCustomMetaDataTypeSet) {
+                MetaData.Custom tribeCustomMetaData = currentState.metaData().custom(changedCustomMetaDataType);
+                MetaData.Custom mergedCustomMetaData = mergedCustomMetaDataMap.get(changedCustomMetaDataType);
+                if (mergedCustomMetaData == null) {
+                    // custom md has been removed
+                    clusterStateChanged = true;
+                    metaData.removeCustom(changedCustomMetaDataType);
+                } else if (mergedCustomMetaData.equals(tribeCustomMetaData) == false) {
```

Comment:
do we really need this extra protection? I mean the type was already in our changed type list?